README.md

```diff
@@ -90,6 +90,35 @@ but from here onwards you can use the `pganalyze` user instead.
 The collector will automatically use the helper methods
 if they exist in the `pganalyze` schema - otherwise data will be fetched directly.
 
+If you use `enable_log_explain`:
+
+```
+CREATE OR REPLACE FUNCTION pganalyze.explain(query text, params text) RETURNS text AS
+$$
+DECLARE
+  prepared_query text;
+  result text;
+BEGIN
+  SELECT regexp_replace(query, ';+\s*\Z', '') INTO prepared_query;
+  -- this is a check to minimize collector access to your data: it ensures that the
+  -- collector cannot piggyback other queries that could exfiltrate data
+  IF prepared_query LIKE '%;%' OR params LIKE '%;%' THEN
+    RAISE EXCEPTION 'cannot run EXPLAIN when query or parameters contain semicolon';
+  END IF;
+
+  IF length(params) > 0 THEN
+    EXECUTE 'PREPARE pganalyze_explain AS ' || prepared_query;
+    EXECUTE 'EXPLAIN (VERBOSE, FORMAT JSON) EXECUTE pganalyze_explain(' || params || ')' INTO STRICT result;
```

Comment:
I've been thinking if there is a way to avoid checking the `params` value for semicolons, because thats where the restriction is going to be most annoying (assuming you're using prepared statements, the query text itself shouldn't contain semicolons - but as discussed a string used in the query very well might).

I think there may be a way to achieve this, if we run `quote_ident` inside this function, and accept a variadic parameter list (instead of doing the escaping on the Go clientside). Its a bit more PL/pgSQL code, but would allow us to lift the restriction on checking semicolons in `params`. What do you think?

Reply:
I was thinking about that, too, and had convinced myself that we couldn't do that because we would be forcing some bogus type information (i.e., claiming the params are all `text`). I've definitely run into issues in the past where `unknown` goes through implicit casts more easily more easily than text.

But I realized we will run into issues in similar cases even with the quote-in-collector approach, and we could explicitly cast these parameters to `unknown` in the new approach to help avoid that as much as possible. We don't know the actual parameter types in our query samples, right?

I think we'll still run into some issues, but this should be rare enough that we can treat it as an edge case:

```
maciek=# prepare add(integer, integer) as select $1 + $2;
PREPARE
maciek=# execute add(1,2);
 ?column?
----------
        3
(1 row)

maciek=# select pganalyze.explain('select $1 + $2', '''1'', ''2''');
ERROR:  operator is not unique: unknown + unknown
LINE 1: PREPARE pganalyze_explain AS select $1 + $2
                                               ^
HINT:  Could not choose a best candidate operator. You might need to add explicit type casts.
QUERY:  PREPARE pganalyze_explain AS select $1 + $2
CONTEXT:  PL/pgSQL function pganalyze.explain(text,text) line 14 at EXECUTE
```

(or with the updated param-array explain helper):

```
maciek=# select pganalyze.explain('select $1 + $2', array['1', '2']::text[]);
ERROR:  operator is not unique: unknown + unknown
LINE 1: PREPARE pganalyze_explain AS select $1 + $2
                                               ^
HINT:  Could not choose a best candidate operator. You might need to add explicit type casts.
QUERY:  PREPARE pganalyze_explain AS select $1 + $2
CONTEXT:  PL/pgSQL function pganalyze.explain(text,text[]) line 15 at EXECUTE
```
I don't think there's really much we can do about these cases. We should probably catch "Could not choose a best candidate operator" cases and special-case the UI error as something like "Sorry, this explain doesn't work in log-based explain right now; try autoexplain if possible".

Also, I was thinking about other situations that could break log-based EXPLAIN, and I think `search_path` is a big one. Not sure there's an easy fix (at least for complicated cases), but curious if you've given that any thought.

Reply:
> We don't know the actual parameter types in our query samples, right?

I think thats correct - interestingly enough, Postgres seems to generally output most parameters as string values (even when the data type is integer)

> I don't think there's really much we can do about these cases. We should probably catch "Could not choose a best candidate operator"

Yeah, I think for now we can just make sure that gets stored as an EXPLAIN error - its unclear how common this problem would really be (AFAIK when a column is involved, e.g. in an expression, Postgres can infer the type of the parameter from the column type)

> Also, I was thinking about other situations that could break log-based EXPLAIN, and I think search_path is a big one. Not sure there's an easy fix (at least for complicated cases), but curious if you've given that any thought.

I don't think we can do anything about search path, unfortunately. I think the assumption we make with the code right now is that the search_path thats active for the collector user when connected to the DB, is the same as that which was used for the original query. When someone sets the search path on a per-connection basis, this would break. If necessary we could provide more functionality here, but its unclear if thats really needed.

As a side note, I also briefly thought whether the recent `search_path` security issues with Postgres are of concern here (where the attacker tricks you to call a function in an attacker-controlled schema, due to references in the query not being fully qualified), but since we strictly pass everything through `EXPLAIN`, this wouldn't be of a concern.

Reply:
Makes sense. Fun fact: I found two `search_path` vulnerabilities in heroku hobby DBs using that mechanism that could give you superuser on the cluster (thus full access to ~2k other hobby DBs).